### ISSUE TITLE
Dictionaries with no sense of self

### ISSUE THREAD

**Issue by copumpkin** [google/jsonnet#220]
I find that most of the time I'm writing jsonnet, having an object bring a magic `self` into scope is a bit of a pain. I usually end up having to define a `local realSelf = self` in the outer scope to retain the `self` that I actually wanted.

Can we have raw dictionaries with no magic `self` in scope for those times we just want to operate on a pure datastructure?

This is _almost_ the opposite of the `rec` keyword in Nix. For context, `rec` brings a dictionary's keys into scope within that dictionary, and the default is to have no access to the equivalent of `self`. I'm not proposing changing the default at this stage, but it would be nice to have a way to opt out.

cc @roconnor, since whenever I mention Nix @sparkprime cc's him anyway 😄

**Comment by sparkprime** [comment]
I usually give them descriptive names so instead of `local realSelf = self` I use `local service = self` or whatever it is.

To be clear, what you're proposing is a syntax sugar:

`{ inner: {{ x: self.y, y: 3 }}, y: 10 }`

desugared via this rule:
`{{ [f]: e, ... }} -> local fresh = self; { [f]: e[fresh / self] }`

to

`{ inner: local fresh = self; { x: fresh.y, y: 3 }, y: 10 }`

which would manifest as:

`{ "inner": { "x": 10, "y": 3 }, "y": 3 }`

**Comment by sparkprime** [comment]
Assuming {{ }} is the syntax, which is probably unwise because of templating languages and their predilection for {{ }}

**Comment by igorpeshansky** [comment]
Note that this is exactly how ES6 arrow functions work (as compared to regular JS functions).

**Comment by sparkprime** [comment]
Technically, yes, but the presence of self binding in an object is more significantly more legitimate / expected than in a closure.

**Comment by igorpeshansky** [comment]
Right. I meant that JS functions have a "this" binding in them (similar to JSonnet objects), which is very inconvenient (one has to use f.bind(this)). Your proposed sugar is similar to what ES6 tries to achieve with ()=> functions, which don't have a "this" binding, and thus capture one from the outer scope.
On an unrelated note, I wonder if you should even call those sugared blocks "objects" anymore.

**Comment by benley** [comment]
I would secretly be pleased if jsonnet used {{}} in its grammar, because it would discourage the use of text templating tools to generate jsonnet code ;-)
